okio/src/main/java/okio/GzipSource.kt

```diff
@@ -0,0 +1,220 @@
+    // need to CRC the entire header. Then we read the magic ID1ID2 sequence.
+    // We can skip everything else in the first 10 bytes.
+    // +---+---+---+---+---+---+---+---+---+---+
+    // |ID1|ID2|CM |FLG|     MTIME     |XFL|OS | (more-->)
+    // +---+---+---+---+---+---+---+---+---+---+
+    source.require(10)
+    val flags = source.buffer().getByte(3).toInt()
+    val fhcrc = flags.getBit(FHCRC)
+    if (fhcrc) updateCrc(source.buffer(), 0, 10)
+
+    val id1id2 = source.readShort()
+    checkEqual("ID1ID2", 0x1f8b, id1id2.toInt())
+    source.skip(8)
+
+    // Skip optional extra fields.
+    // +---+---+=================================+
+    // | XLEN  |...XLEN bytes of "extra field"...| (more-->)
+    // +---+---+=================================+
+    if (flags.getBit(FEXTRA)) {
+      source.require(2)
+      if (fhcrc) updateCrc(source.buffer(), 0, 2)
+      val xlen = source.buffer().readShortLe().toLong()
+      source.require(xlen)
+      if (fhcrc) updateCrc(source.buffer(), 0, xlen)
+      source.skip(xlen)
+    }
+
+    // Skip an optional 0-terminated name.
+    // +=========================================+
+    // |...original file name, zero-terminated...| (more-->)
+    // +=========================================+
+    if (flags.getBit(FNAME)) {
+      val index = source.indexOf(0)
+      if (index == -1L) throw EOFException()
+      if (fhcrc) updateCrc(source.buffer(), 0, index + 1)
+      source.skip(index + 1)
+    }
+
+    // Skip an optional 0-terminated comment.
+    // +===================================+
+    // |...file comment, zero-terminated...| (more-->)
+    // +===================================+
+    if (flags.getBit(FCOMMENT)) {
+      val index = source.indexOf(0)
+      if (index == -1L) throw EOFException()
+      if (fhcrc) updateCrc(source.buffer(), 0, index + 1)
+      source.skip(index + 1)
+    }
+
+    // Confirm the optional header CRC.
+    // +---+---+
+    // | CRC16 |
+    // +---+---+
+    if (fhcrc) {
+      checkEqual("FHCRC", source.readShortLe().toInt(), crc.value.toShort().toInt())
+      crc.reset()
+    }
+  }
+
+  @Throws(IOException::class)
+  private fun consumeTrailer() {
+    // Read the eight-byte trailer. Confirm the body's CRC and size.
+    // +---+---+---+---+---+---+---+---+
+    // |     CRC32     |     ISIZE     |
+    // +---+---+---+---+---+---+---+---+
+    checkEqual("CRC", source.readIntLe(), crc.value.toInt())
+    checkEqual("ISIZE", source.readIntLe(), inflater.bytesWritten.toInt())
+  }
+
+  override fun timeout(): Timeout = source.timeout()
+
+  @Throws(IOException::class)
+  override fun close() = inflaterSource.close()
+
+  /** Updates the CRC with the given bytes.  */
+  private fun updateCrc(buffer: Buffer, offset: Long, byteCount: Long) {
+    var offset = offset
+    var byteCount = byteCount
+    // Skip segments that we aren't checksumming.
+    var s = buffer.head
```

Comment:
nit: I think this wants to be

```
var s = buffer.head!!
while (offset >= s.limit - s.pos) {
  offset -= s.limit - s.pos
  s = s.next!!
}
```

ie. it’s non-null on assignment, not on use

Reply:
Ah, yep. I only did a quick pass. This is mostly auto-converted code.